File: lc_soln/0019_remove_nth_node_from_end_of_list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* fast = head;
        for (int i = 0; i < n; ++i) {
            fast = fast->next;
        }
        ListNode* prev = nullptr;
        ListNode* curr = head;
        while (fast != nullptr) {
            fast = fast->next;
            prev = curr;
            curr = curr->next;
        }
        if (curr == head) {
            return head->next;
        }
        prev->next = curr->next;
        return head;
    }
};

int main() {
    return 0;
}

File: lc_soln/0019_remove_nth_node_from_end_of_list/solution.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution
{
    public ListNode RemoveNthFromEnd(ListNode head, int n)
    {
        var fast = head;
        while (n > 0)
        {
            fast = fast.next;
            n--;
        }
        ListNode prev = null;
        ListNode curr = head;
        while (fast is not null)
        {
            fast = fast.next;
            prev = curr;
            curr = curr.next;
        }
        if (curr == head)
        {
            return head.next;
        }
        prev.next = curr.next;
        return head;
    }
}
